File: examples/otel-setup.js
const { NodeSDK } = require('@opentelemetry/sdk-node');
const { OTLPTraceExporter } = require('@opentelemetry/exporter-trace-otlp-grpc');
const { getNodeAutoInstrumentations } = require('@opentelemetry/auto-instrumentations-node');
const { Resource } = require('@opentelemetry/resources');
const { SemanticResourceAttributes } = require('@opentelemetry/semantic-conventions');

const sdk = new NodeSDK({
  resource: new Resource({
    [SemanticResourceAttributes.SERVICE_NAME]: 'order-service', // fallback component_id
  }),
  traceExporter: new OTLPTraceExporter({
    url: process.env.OTEL_EXPORTER_OTLP_ENDPOINT || 'http://localhost:4317',
  }),
  instrumentations: [getNodeAutoInstrumentations()],
});

sdk.start();
process.on('SIGTERM', () => sdk.shutdown());

// Usage Example
const { trace, SpanStatusCode } = require('@opentelemetry/api');
const tracer = trace.getTracer('order-service', '1.0.0');

async function processOrder(orderId, workflowId) {
  return tracer.startActiveSpan('process-order', async (span) => {
    try {
      span.setAttribute('workflow.id',  workflowId);
      span.setAttribute('component.id', 'order-service');
      span.setAttribute('entity.id',    orderId); // powers TraceTimelinePanel

      const result = await doWork(orderId);
      span.setStatus({ code: SpanStatusCode.OK });
      return result;
    } catch (err) {
      span.setStatus({ code: SpanStatusCode.ERROR, message: err.message });
      span.recordException(err);
      throw err;
    } finally {
      span.end(); // duration_ms is calculated automatically from this
    }
  });
}

async function doWork(orderId) {
  return new Promise(resolve => setTimeout(() => resolve(`Processed ${orderId}`), 100));
}

File: examples/otel-setup.py
# pip install opentelemetry-sdk opentelemetry-exporter-otlp-proto-grpc

from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
import time

resource = Resource.create({"service.name": "payment-service"})
provider = TracerProvider(resource=resource)
provider.add_span_processor(
    BatchSpanProcessor(OTLPSpanExporter(endpoint="http://localhost:4317"))
)
trace.set_tracer_provider(provider)
tracer = trace.get_tracer(__name__)

def process_payment(payment_id: str, workflow_id: str):
    with tracer.start_as_current_span("process-payment") as span:
        try:
            span.set_attribute("workflow.id",  workflow_id)
            span.set_attribute("component.id", "payment-service")
            span.set_attribute("entity.id",    payment_id)
            
            # ... business logic ...
            time.sleep(0.1)
            
            span.set_status(trace.StatusCode.OK)
        except Exception as e:
            span.record_exception(e)
            span.set_status(trace.StatusCode.ERROR, str(e))
            raise
